### ISSUE TITLE
partner can't delete users

### ISSUE THREAD

**Issue by ClaireValdivia** [usdigitalresponse/usdr-gost#2202]
### Why is this issue important?

Mel Evans from City of Sparks reached out trying to delete two users and she's not able to. Logged in and tried to do so as well and was not able to - delete button does nothing

### Current State

Delete button on User page does not remove user. Aditya confirmed that any users who have a saved search can't be deleted.

The error trace can be found [here in Datadog](https://app.datadoghq.com/apm/trace/1488372201979358490?colorBy=service&env=production&graphType=flamegraph&shouldShowLegend=true&sort=time&spanID=1488372201979358490&spanViewType=errors&timeHint=1699541585263.001&view=spans).

```
error: delete from "users" where "id" = $1 - update or delete on table "users" violates foreign key constraint "grants_saved_searches_created_by_foreign" on table "grants_saved_searches"
    at Parser.parseErrorMessage (/app/node_modules/pg-protocol/dist/parser.js:287:98)
    at Parser.handlePacket (/app/node_modules/pg-protocol/dist/parser.js:126:29)
    at Parser.parse (/app/node_modules/pg-protocol/dist/parser.js:39:38)
    at TLSSocket.<anonymous> (/app/node_modules/pg-protocol/dist/index.js:11:42)
    at TLSSocket.emit (node:events:520:28)
    at addChunk (node:internal/streams/readable:315:12)
    at readableAddChunk (node:internal/streams/readable:289:9)
    at TLSSocket.Readable.push (node:internal/streams/readable:228:10)
    at TLSWrap.onStreamRead (node:internal/stream_base_commons:190:23)
    at TLSWrap.callbackTrampoline (node:internal/async_hooks:130:17)
```

As shown above, the cause of this issue is the existence of a foreign key constraint, `grants_saved_searches_created_by_foreign` on table `grants_saved_searches`, which links the `grants_saved_searches.created_by` column to the `users.id` column. Therefore, when user with one or more saved searches is being deleted, the attempt fails (due to a violation of the foreign key constraint) because it would "orphan" the `grants_saved_searches` rows associated with that user.

### Expected State

- Delete button should remove user regardless of whether they have a saved search.
- When the user being deleted has created saved searches, those saved searches are also deleted.

### Implementation Plan

As there is limited value in saved searches created by deleted users, this issue should be resolved by implementing `ON DELETE CASCADE` behavior on the foreign key constraint. Once implemented, deleting a row from the `users` table will automatically also delete any row(s) in the `grants_saved_searches` table where the `created_by` column of the latter references the `id` column of the former.

The expected behavior should be implemented through one or more Knex migrations. The following steps must be implemented in separate transactions (either by creating separate, subsequent Knex migrations or by instrumenting multiple transactions within a single Knex migration) in order to avoid downtime when running the migration:
1. Rename the existing `grants_saved_searches_created_by_foreign` constraint (e.g. to `grants_saved_searches_created_by_foreign_old`)
    ```sql
    alter table grants_saved_searches 
    rename CONSTRAINT grants_saved_searches_created_by_foreign 
    TO grants_saved_searches_created_by_foreign_old;
    ```
2. Create the new constraint **without validation**:
    ```sql
    alter table grants_saved_searches
    add constraint grants_saved_searches_created_by_foreign
    foreign key (created_by) references users(id)
    on delete cascade
    not valid;
    ```
    Note: The `NOT VALID` marker prevents Postgres from locking out concurrent updates during creation of the constraint. See [these notes](https://www.postgresql.org/docs/13/sql-altertable.html#SQL-ALTERTABLE-NOTES) in the Postgres 13 documentation for more details.
3. Drop the old foreign key constraint:
    ```sql
    alter table grants_saved_searches
    drop constraint grants_saved_searches_created_by_foreign_old;
    ```
4. Finally, validate the new constraint:
    ```sql
    alter table grants_saved_searches
    validate constraint grants_saved_searches_created_by_foreign;
    ```
    Note: Upon completion, the `NOT VALID` marker will be removed from the new constraint in Postgres, and will only require row-level locking.

### Relevant Code Snippets

Current foreign-key constraint definition (before Step 1):
```sql
"grants_saved_searches_created_by_foreign" FOREIGN KEY (created_by) REFERENCES users(id)
```

Target foreign-key constraint definition (after Step 4):
```sql
"grants_saved_searches_created_by_foreign" FOREIGN KEY (created_by) REFERENCES users(id) ON DELETE CASCADE
```

**Comment by TylerHendrickson** [comment]
Closing as duplicate of #2176

**Comment by TylerHendrickson** [comment]
@ClaireValdivia FYI Added implementation details.

**Comment by ClaireValdivia** [comment]
@dysmento I just tested this in Grant Finder staging and still can't delete a user who has a saved search. steps I took:
- logged into staging as an admin (cvaldivia@usdigitalrepsonse.org)
-  created user "cvaldivia+testdelete@usdigitalresponse.org"
- logged in as the new user and created a saved search
- logged back in as admin (cvaldivia@usdigitalrepsonse.org) and tried to delete cvaldivia+testdelete@usdigitalresponse.org
- nothing happens when I click the delete button; I tried refreshing the page, logging out and back in, but can't delete the user

**Comment by ClaireValdivia** [comment]
tested and confirmed that I can delete - deleted cvaldivia+test7@usdigitalresponse.org, which had one saved search
